{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import qiskit\n",
    "from qiskit import IBMQ\n",
    "from qiskit import Aer\n",
    "from qiskit import QuantumCircuit, ClassicalRegister,QuantumRegister, QiskitError\n",
    "from qiskit.tools.visualization import circuit_drawer\n",
    "from qiskit.tools.visualization import plot_histogram\n",
    "from qiskit.providers.aer import noise\n",
    "import random\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import collections as c\n",
    "import math\n",
    "from math import *\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Defining some auxiliary functions\n",
    "def convertToNumEncoding(genome_str):\n",
    "    bin_str = \"\"\n",
    "    for i in range(0, len(genome_str)):\n",
    "        if genome_str[i] == 'A':\n",
    "            bin_str = bin_str + \"0\"\n",
    "        elif genome_str[i] == 'C':\n",
    "            bin_str = bin_str + \"1\"\n",
    "        elif genome_str[i] == 'G':\n",
    "            bin_str = bin_str + \"2\"\n",
    "        elif genome_str[i] == 'T':\n",
    "            bin_str = bin_str + \"3\"\n",
    "    return bin_str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length: 1709\n"
     ]
    }
   ],
   "source": [
    "## Initializing global variables\n",
    "# Alphabet set (0, 1, 2, 3) := {A, C, G, T} for DNA Nucleotide bases\n",
    "AS = {'00', '01', '10', '11'}\n",
    "A = len(AS)\n",
    "genome_file = open(\"Sars_cov_2.txt\", \"r\")\n",
    "#genome_file = open(\"Carsonella_rudii.txt\", \"r\")\n",
    "#genome_file = open(\"nsp14.txt\", \"r\")\n",
    "RG = genome_file.read()\n",
    "LenRG=len(RG)\n",
    "print('Length:', LenRG)\n",
    "genome_file.close()\n",
    "RG = convertToNumEncoding(RG)\n",
    "RG = RG[0:LenRG]\n",
    "N = len(RG)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Short Read search string\n",
    "SR = '23?'\n",
    "M = len(SR)\n",
    "Q_A = ceil(log2(A))\n",
    "# Data Qubits\n",
    "Q_D = Q_A * M\n",
    "# Tag Qubits\n",
    "Q_T = ceil(log2(N - M + 1))\n",
    "# Ancilla Qubits\n",
    "Q_anc = 8\n",
    "# Ancilla qubits ids\n",
    "anc = []\n",
    "for qi in range(0, Q_anc):\n",
    "    anc.append(Q_D + Q_T + qi)\n",
    "# Total number of qubits\n",
    "Q = Q_D + Q_T + Q_anc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Start codon\n",
    "Start = '032'\n",
    "Xl = len(Start)\n",
    "\n",
    "#Stop codon\n",
    "Stop = ['302','320','300']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "IBMQ.save_account('APIKEY', overwrite=True)\n",
    "## Initialization of IBM QX\n",
    "IBMQ.enable_account('5b3b5238d1fec1257d958a113fb4936b7cb56a3bd4c1522b5b8c13366faca29feb421a44c3fb07bcf747db804e0e31a5f64694bcf54ba2c1e25310615110bc54')\n",
    "#IBMQ.enable_account('bf333c31a0a3b9e5829fa0a0e0097216062f044a366ff7cce49dc0f004ad35da0d8067cb98c1f59e8d79704b055dac79de9e88c4e10306dc2cb14e20183f4e68')\n",
    "#provider = IBMQ.get_provider()\n",
    "#backend = provider.get_backend('ibmq_qasm_simulator')\n",
    "backend = Aer.get_backend('qasm_simulator')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Defining functions\n",
    "def QAM():\n",
    "    print('Reference genome:', RG, 'Sars_cov_2', 'Length of the genome:', LenRG)\n",
    "    #print('Carsonella rudii:', RG1)\n",
    "    #print('Chosen pattern for testing:', SR)\n",
    "    print('Total number of qubits:', Q)\n",
    "    print('Number of ancilla qubits:', Q_anc)\n",
    "    qr = qiskit.QuantumRegister(Q)\n",
    "    cr = qiskit.ClassicalRegister(Q_T)\n",
    "    qc = qiskit.QuantumCircuit(qr, cr)\n",
    "    # Patterns are stored\n",
    "    generateInitialState(qc, qr)\n",
    "# Patterns are turned into Hamming Distances\n",
    "    evolveToHammingDistances(qc, qr)\n",
    "# Marking the zero Hamming Distance states\n",
    "    markZeroHammingDistance(qc, qr)\n",
    "    inversionAboutMean(qc, qr)\n",
    "# Applying again this funcions turns back the Hamming Distances into the original patterns\n",
    "    evolveToHammingDistances(qc, qr)\n",
    "# Marking the stored patterns\n",
    "    markStoredPatterns(qc, qr)\n",
    "    # From patterns to Hamming Distances again\n",
    "    evolveToHammingDistances(qc, qr)\n",
    "    inversionAboutMean(qc, qr)\n",
    "    # Grover’s iterations\n",
    "    it = 0\n",
    "    for i in range(0, 1):\n",
    "        markZeroHammingDistance(qc, qr)\n",
    "        inversionAboutMean(qc, qr)\n",
    "        it = it + 1\n",
    "    print(\"Grover’s algorithm had {} iterations.\".format(int(it)))\n",
    "    finalGroverMeasurement(qc, qr, cr)\n",
    "    QAM2(qc, qr)\n",
    "    return qc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialization as suggested by L. C. L. Hollenberg\n",
    "def generateInitialState(qc, qr):\n",
    "    for qi in range(0, Q_T):\n",
    "        qc.h(qr[qi])\n",
    "    \n",
    "    control_qubits = []\n",
    "    for ci in range(0, Q_T):\n",
    "        control_qubits.append(qr[ci])\n",
    "    \n",
    "    ancilla_qubits = []\n",
    "    for qi in anc:\n",
    "        ancilla_qubits.append(qr[qi])\n",
    "    \n",
    "    for qi in range(0, N - M + 1):\n",
    "        qis = format(qi, '0' + str(Q_T) + 'b')\n",
    "        for qisi in range(0, Q_T):\n",
    "            if qis[qisi] == '0':\n",
    "                qc.x(qr[qisi])\n",
    "        wMi = RG[qi:qi + M]\n",
    "        #print(\"Tag: {} - Data: {}\".format(qis, wMi))\n",
    "\n",
    "        for wisi in range(0, M):\n",
    "            wisia = format(int(wMi[wisi]), '0' + str(Q_A) + 'b')\n",
    "            for wisiai in range(0, Q_A):\n",
    "                if wisia[wisiai] == '1':\n",
    "                    qc.mct(control_qubits, qr[Q_T + wisi * Q_A + wisiai], ancilla_qubits)\n",
    "        for qisi in range(0,Q_T):\n",
    "            if qis[qisi] == '0':\n",
    "                qc.x(qr[qisi])\n",
    "                \n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def QAM2(qc, qr):\n",
    "    ## To compare given search string (GD)\n",
    "    control_qubits = []\n",
    "    for qi in range(1, Q_D + Q_T):\n",
    "        control_qubits.append(qr[qi])\n",
    "    \n",
    "    ancilla_qubits = []\n",
    "    for qi in anc:\n",
    "        ancilla_qubits.append(qr[qi])\n",
    "    Mark = []\n",
    "    Mark1= []\n",
    "    for qi in range(0, N - M + 1):\n",
    "        #print(\"1\", qi)\n",
    "        Seq = []\n",
    "        \n",
    "        qis = format(qi, '0' + str(Q_T) + 'b')\n",
    "        for qisi in range(0, Q_T):\n",
    "            if qis[qisi] == '0':\n",
    "                qc.x(qr[qisi])\n",
    "        wMi = RG[qi:qi + M]\n",
    "        #print(\"1\",wMi)\n",
    "        counter = compare(wMi, Start)\n",
    "        #print(counter, \"yes\")\n",
    "        \n",
    "        if counter == 3 :\n",
    "            #print(Start, \"found at \", qis)\n",
    "            y = int(qis, 2)\n",
    "            \n",
    "            #print(\"y\",y)\n",
    "            \n",
    "            for qi in range(y, N - M + 1, 3):\n",
    "                qis = format(qi, '0' + str(Q_T) + 'b')\n",
    "                for qisi in range(0, Q_T):\n",
    "                    if qis[qisi] == '0':\n",
    "                        qc.x(qr[qisi])\n",
    "                wmi = RG[qi:qi + M]\n",
    "                \n",
    "                #print(\"Tag: {} - Data: {}\".format(qis, wmi))\n",
    "                Mark.append(int(qis,2))\n",
    "                for i in range(0,3):\n",
    "                    StopC = compare(wmi, Stop[i])\n",
    "                    if StopC == 3:\n",
    "                        break\n",
    "                Seq.append(wmi)\n",
    "                if StopC == 3:\n",
    "                    break\n",
    "                \n",
    "            z=int(qis,2)+3\n",
    "            Length=len(Seq)\n",
    "            if Length > 30:    \n",
    "                print(Length, \"\\t codons long ORF found at \\t\", y, \":\\t\",y,\"-\",z)\n",
    "##                \n",
    "##    plt.plot(Mark[:],'ro')\n",
    "##    \n",
    "##    #plt.gca().invert_yaxis()\n",
    "##    plt.xlabel('genome sequence')\n",
    "##    plt.show()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def compare(a,b):\n",
    "    #print(a,b)\n",
    "    c = 0\n",
    "    for x, y in zip(a, b):\n",
    "        if x == y:\n",
    "            c+=1\n",
    "            #print(x,y)\n",
    "    return c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "def evolveToHammingDistances(qc, qr):\n",
    "    for pi in range(0, M):\n",
    "        if SR[pi] == '?':\n",
    "            continue\n",
    "        ppi = format(int(SR[pi]), '0' + str(Q_A) + 'b')\n",
    "        for ppii in range(0, Q_A):\n",
    "            if ppi[ppii] == '1':\n",
    "                qc.x(qr[Q_T + pi * Q_A + ppii])\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Oracle to mark zero Hamming Distance\n",
    "def markZeroHammingDistance(qc, qr):\n",
    "    for qi in range(0, Q_D):\n",
    "        qc.x(qr[Q_T + qi])\n",
    "    control_qubits = []\n",
    "    for mi in range(0, M):\n",
    "        if SR[mi] == '?':\n",
    "            for ai in range(0, Q_A):\n",
    "                control_qubits.append(qr[Q_T + mi * Q_A + ai])\n",
    "    \n",
    "    ancilla_qubits = []\n",
    "    for qi in anc:\n",
    "        ancilla_qubits.append(qr[qi])\n",
    "    target = control_qubits.pop()\n",
    "    qc.h(target)\n",
    "    qc.mct(control_qubits, target, ancilla_qubits)\n",
    "    qc.h(target)\n",
    "    for qi in range(0, Q_D):\n",
    "        qc.x(qr[Q_T + qi])\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# Inversion about mean\n",
    "def inversionAboutMean(qc, qr):\n",
    "    for si in range(0, Q_D + Q_T):\n",
    "        qc.h(qr[si])\n",
    "        qc.x(qr[si])\n",
    "    \n",
    "    control_qubits = []\n",
    "    for qi in range(1, Q_D + Q_T):\n",
    "        control_qubits.append(qr[qi])\n",
    "    \n",
    "    ancilla_qubits = []\n",
    "    for qi in anc:\n",
    "        ancilla_qubits.append(qr[qi])\n",
    "    qc.h(qr[0])\n",
    "    qc.mct(control_qubits, qr[0], ancilla_qubits)\n",
    "    qc.h(qr[0])\n",
    "    for si in range(0, Q_D + Q_T):\n",
    "        qc.x(qr[si])\n",
    "        qc.h(qr[si])\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Oracle to mark stored patterns\n",
    "def markStoredPatterns(qc, qr):\n",
    "    control_qubits = []\n",
    "    for qi in range(1, Q_D + Q_T):\n",
    "        control_qubits.append(qr[qi])\n",
    "    \n",
    "    ancilla_qubits = []\n",
    "    for qi in anc:\n",
    "        ancilla_qubits.append(qr[qi])\n",
    "   \n",
    "    for qi in range(0, N - M + 1):\n",
    "        qis = format(qi, '0' + str(Q_T) + 'b')\n",
    "        for qisi in range(0, Q_T):\n",
    "            if qis[qisi] == '0':\n",
    "                qc.x(qr[qisi])\n",
    "        wMi = RG[qi:qi + M]\n",
    "        for wisi in range(0, M):\n",
    "            wisia = format(int(wMi[wisi]), '0' + str(Q_A) + 'b')\n",
    "            for wisiai in range(0, Q_A):\n",
    "                if wisia[wisiai] == '0':\n",
    "                    qc.x(qr[Q_T + wisi * Q_A + wisiai])\n",
    "        qc.h(qr[0])\n",
    "        qc.mct(control_qubits, qr[0], ancilla_qubits)\n",
    "        qc.h(qr[0])\n",
    "        for wisi in range(0,M):\n",
    "            wisia = format(int(wMi[wisi]), '0' + str(Q_A) + 'b')\n",
    "            for wisiai in range(0, Q_A):\n",
    "                if wisia[wisiai] == '0':\n",
    "                    qc.x(qr[Q_T + wisi * Q_A + wisiai])\n",
    "        for qisi in range(0, Q_T):\n",
    "            if qis[qisi] == '0':\n",
    "                qc.x(qr[qisi])\n",
    "    return\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Final measurement\n",
    "def finalGroverMeasurement(qc, qr, cr):\n",
    "    for qi in range(0, Q_T):\n",
    "        qc.measure(qr[qi], cr[qi])\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reference genome: 02132000032300102201313330002033230230022300310132223301031130101022101130101011310232332010130003310000132002233303232332010301132210301130022010320113030200201310313130320322233330000320033031002330032233011130010323330310111212002002130300201032301232103220332213312032312022223231032130130202002132332230110033301133301021302233333130102232330011302332132301130102233032332030101130030030102033333110202330232130001101121132202031003330001011310301101330323010002201331133220032302321230300020332301000323300232010101330000031313132010202312303332313303222101032213332023320103130320023033332320000302201132021210113233231303232030201232110103213333110132133102010133032113233221031033130332203332033012313030031123330320332032331001003222233330102230011301000210011032031323033231002311032230032101032302130233232032100310320130223231302132311012023213332330021232332013220130332003031130300332232032001320020330032122133230200022331001010322332330002132103303302102010003311102331331012010332230011130002130330023232301131002132032302003220023313032032101021133230232010002133030000302002003303313033133032110101033132010003310102032232303211303333220033210032312030203031132130033110332333230203332010130202321303130011330013321132233232032232210233323032300030001032103311010101102133332030000232133332330033300001003301103333313033013132010231103232023131032200001002302323102030302033032301101300023132130123230300101233210033302232232132313230201031032130032023010203323031312032133030010320320313102132213330213323222333010001003332030133030011313220010133330100201331020233302000032322133330032332300030022201013332032201001022232002301102333130310330030010132333010100002332032232332032302003323 Sars_cov_2 Length of the genome: 1709\n",
      "Total number of qubits: 25\n",
      "Number of ancilla qubits: 8\n",
      "Grover’s algorithm had 1 iterations.\n",
      "521 \t codons long ORF found at \t 145 :\t 145 - 1708\n",
      "513 \t codons long ORF found at \t 169 :\t 169 - 1708\n",
      "512 \t codons long ORF found at \t 172 :\t 172 - 1708\n",
      "508 \t codons long ORF found at \t 184 :\t 184 - 1708\n",
      "498 \t codons long ORF found at \t 214 :\t 214 - 1708\n",
      "38 \t codons long ORF found at \t 371 :\t 371 - 485\n",
      "417 \t codons long ORF found at \t 457 :\t 457 - 1708\n",
      "401 \t codons long ORF found at \t 505 :\t 505 - 1708\n",
      "375 \t codons long ORF found at \t 583 :\t 583 - 1708\n",
      "329 \t codons long ORF found at \t 721 :\t 721 - 1708\n",
      "294 \t codons long ORF found at \t 826 :\t 826 - 1708\n",
      "255 \t codons long ORF found at \t 943 :\t 943 - 1708\n",
      "52 \t codons long ORF found at \t 1109 :\t 1109 - 1265\n",
      "43 \t codons long ORF found at \t 1136 :\t 1136 - 1265\n",
      "34 \t codons long ORF found at \t 1163 :\t 1163 - 1265\n",
      "70 \t codons long ORF found at \t 1498 :\t 1498 - 1708\n",
      "69 \t codons long ORF found at \t 1501 :\t 1501 - 1708\n",
      "Circuit depth: 19761\n",
      "Number of gates: 99491\n"
     ]
    }
   ],
   "source": [
    "# Main function\n",
    "if __name__ == '__main__':\n",
    "# Printing some data for testing\n",
    "    qc = QAM()\n",
    "    print(\"Circuit depth: {}\".format(qc.depth()))\n",
    "# Total number of gates\n",
    "    print(\"Number of gates: {}\".format(len(qc.data)))\n",
    "    gate_num = 1\n",
    "    for item in qc.data:\n",
    "        qb_list = \"\"\n",
    "        for qb in item[1]:\n",
    "            qb_list = qb_list + str(qb.index) + \", \"\n",
    "        qb_list = qb_list[:len(qb_list)-2]\n",
    "        #print(\"#{}: {}, {}\".format(gate_num, item[0].name, qb_list))\n",
    "        gate_num = gate_num + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
